File: pippip-native-ios/API/NewAccountResponse.m
//
//  NewAccountResponse.m
//  pippip-native-ios
//
//  Created by Steve Brenneis on 10/10/17.
//  Copyright © 2017 seComm. All rights reserved.
//

#import "NewAccountResponse.h"
#import "pippip_native_ios-Swift.h"
#import "CKRSACodec.h"

@interface NewAccountResponse ()
{
    SessionState *sessionState;
}
@end

@implementation NewAccountResponse

- (instancetype)init {
    self = [super init];

    sessionState = [[SessionState alloc] init];
    return self;

}

- (BOOL)processResponse:(NSDictionary *)response errorDelegate:(id<ErrorDelegate>)errorDelegate {

    NSString *dataStr = [response objectForKey:@"data"];
    NSString *errorStr = [response objectForKey:@"error"];
    if (errorStr != nil) {
        [errorDelegate responseError:errorStr];
        return NO;
    }
    if (dataStr == nil) {
        [errorDelegate responseError:@"Invalid server response"];
        return NO;
    }
    else {
        NSData *data = [[NSData alloc] initWithBase64EncodedString:dataStr options:0];
        if (data == nil) {
            [errorDelegate responseError:@"Response encoding error"];
            return NO;
        }
        else {
            NSError *error;
            CKRSACodec *codec = [[CKRSACodec alloc] initWithData:data];
            [codec decrypt:sessionState.userPrivateKey withError:&error];
            if (error != nil) {
                [errorDelegate responseError:[error localizedDescription]];
                return NO;
            }
            sessionState.accountRandom = [codec getBlock];
            return YES;
        }
    }

}

@end

File: pippip-native-ios/API/NewAccountResponse.swift
//
//  NewAccountResponse.swift
//  pippip-native-ios
//
//  Created by Steve Brenneis on 5/30/18.
//  Copyright © 2018 seComm. All rights reserved.
//

import UIKit
import ObjectMapper
import CocoaLumberjack

class NewAccountResponse: NSObject, APIResponseProtocol {
    
    var error: String?
    var sessionId: Int32?
    var authToken: Int64?
    var data: String?
    var postId: Int = 0

    var alertPresenter = AlertPresenter()
    var sessionState = SessionState.instance
    
    required init?(map: Map) {
        if map.JSON["sessionId"] == nil {
            return nil
        }
        if map.JSON["error"] == nil &&  map.JSON["data"] == nil {
            return nil
        }
    }

    func mapping(map: Map) {

        sessionId <- map["sessionId"]
        error <- map["error"]
        data <- map["data"]
        
    }
    
    func processResponse() throws {

        if let encoded = Data(base64Encoded: data!) {
            do {
                let codec = CKRSACodec(data: encoded)
                try codec.decrypt(sessionState.userPrivateKey!)
                sessionState.accountRandom = codec.getBlock()
            }
            catch {
                DDLogError("Error decrypting new account response: \(error.localizedDescription)")
                throw ServerResponseError.invalidServerResponse
            }
        }
        else {
            throw ServerResponseError.invalidResponseEncoding
        }
        
    }

}
